llama_index/query_pipeline/query.py

```diff
@@ -0,0 +1,582 @@
+    module_inputs: Dict[str, Any],
+) -> None:
+    """Add input to module deps inputs."""
+    # get relevant output from link
+    if link.src_key is None:
+        # ensure that output_dict only has one key
+        if len(output_dict) != 1:
+            raise ValueError("Output dict must have exactly one key.")
+        output = next(iter(output_dict.values()))
+    else:
+        output = output_dict[link.src_key]
+
+    # now attach output to relevant input key for module
+    if link.dest_key is None:
+        free_keys = module.free_req_input_keys
+        # ensure that there is only one remaining key given partials
+        if len(free_keys) != 1:
+            raise ValueError(
+                "Module input keys must have exactly one key if "
+                "dest_key is not specified. Remaining keys: "
+                f"in module: {free_keys}"
+            )
+        module_inputs[next(iter(free_keys))] = output
+    else:
+        module_inputs[link.dest_key] = output
+
+
+def print_debug_input(
+    module_key: str,
+    input: Dict[str, Any],
+    val_str_len: int = 200,
+) -> None:
+    """Print debug input."""
+    output = f"> Running module {module_key} with input: \n"
+    for key, value in input.items():
+        # stringify and truncate output
+        val_str = (
+            str(value)[:val_str_len] + "..."
+            if len(str(value)) > val_str_len
+            else str(value)
+        )
+        output += f"{key}: {val_str}\n"
+
+    print_text(output + "\n", color="llama_lavender")
+
+
+class QueryPipeline(QueryComponent):
+    """A query pipeline that can allow arbitrary chaining of different modules.
+
+    A pipeline itself is a query component, and can be used as a module in another pipeline.
+
+    """
+
+    callback_manager: CallbackManager = Field(
+        default_factory=lambda: CallbackManager([]), exclude=True
+    )
+
+    module_dict: Dict[str, QueryComponent] = Field(
+        default_factory=dict, description="The modules in the pipeline."
+    )
+    edge_dict: Dict[str, List[Link]] = Field(
+        default_factory=dict, description="The edges in the pipeline."
+    )
+    verbose: bool = Field(
+        default=False, description="Whether to print intermediate steps."
+    )
+
+    # root_keys: List[str] = Field(
+    #     default_factory=list, description="The keys of the root modules."
+    # )
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    def __init__(
+        self,
+        callback_manager: Optional[CallbackManager] = None,
+        chain: Optional[Sequence[QUERY_COMPONENT_TYPE]] = None,
+        **kwargs: Any,
+    ):
+        super().__init__(
+            callback_manager=callback_manager or CallbackManager([]),
+            **kwargs,
+        )
+
+        if chain is not None:
+            # generate implicit link between each item, add
+            self.add_chain(chain)
+
+    def add_chain(self, chain: Sequence[QUERY_COMPONENT_TYPE]) -> None:
```

Comment:
should enforce existing pipeline is empty. otherwise need additional linkage to tail of existing pipeline.